### PR TITLE
Add PR links to release post template output, and allow blog ID to be specified.

### LINKED ISSUE #34248: Code Analyzer: Add PR links in change descriptions
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**Issue Description:**

In going through previous final release posts, it appears we used to include relevant PR numbers when highlighting important changes. These weren't included in Code Analyzer's original spec because it was a CLI tool, but since the output is used to automatically generate release posts now, we should include PR numbers in the output.

image

This issue has two parts:

- [ ] Enhance Code Analyzer to output PR links
- [ ] Update developer blog post creation tools to output resulting PRs number from Code Analyzer

### PR DESCRIPTION
**Pull request by jonathansadowski**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR updates the release post tool such that it's possible to edit previous release posts. It also adds support for specifying a blog ID, if you'd like to test the tool with a different blog. Finally, it links to the PRs that introduced the changes in the templates section.

Closes #34248.

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

1. Make sure Docker is running and in a state to run `wp-env`
1. Checkout this branch
1. `cd` to the `tools/release-posts` directory
1. Follow the [instructions in the readme](https://github.com/woocommerce/woocommerce/blob/trunk/tools/release-posts/README.md#publishing-draft-posts) to setup for publishing release posts. If you're doing this with a test blog, use that site ID when granting access.
1. The tool will open a browser window, so be sure before running that your focused browser is logged in with access to the blog you'll use for testing.
1. Run the following: `GITHUB_ACCESS_TOKEN="..." WPCOM_OAUTH_CLIENT_ID="..." WPCOM_OAUTH_CLIENT_SECRET="..." pnpm run release-post 7.2.0 7.1.0 --siteId=...`  Be sure to provide th values for the missing variables (...). You may also choose to specify the values via your `.env` file.
1. Follow the prompts in the open browser tool to grant permissions to the site corresponding to the `siteId` that you specified.
1. Verify that the post that is created contains a link to the PR(s) that introduced the template changes.
1. Create a new post on your test site manually, add `<!-- release:templates --> This will be replaced <!-- /release:templates -->` to the source of the post you created. Note the post id.
1. Re-run the previous command with `--editPost=...`, specifying the ID of the post you just created.
1. Verify that that section of the post gets updated properly.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by psealock** [comment]
I keep running into this error. Does it look familiar?

```
GitError: Cloning into '/var/folders/2p/zg7bwl6d4mzb1w5m42p48yn80000gn/T/woocommerce'...
error: RPC failed; curl 56 LibreSSL SSL_read: SSL_ERROR_SYSCALL, errno 54
error: 3218 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
```
image

**Comment by jonathansadowski** [comment]
> I keep running into this error. Does it look familiar?

Hmm. No, that one doesn't. Do you by-chance get the same error when running the command prior to these changes with options other than `--outputOnly`, or does that work for you?

**Comment by psealock** [comment]
I'm having the same issues with just this command `pnpm run release-post contributors -- "7.2.0" "7.1.0"`

I'll need to try again with a fresh install

File: tools/code-analyzer/src/lib/scan-changes.ts
Comment (samueljseay) [inline tools/code-analyzer/src/lib/scan-changes.ts]: This is not type safe and TS complains. I think everything done here could be wrapped into the const declaration and it would be typesafe:

You don't need to worry about checking strictly for undefined.

```suggestion
	Logger.startTask( `Making temporary clone of ${ source }...` );

	const tmpRepoPath =
		typeof clonedPath !== 'undefined'
			? clonedPath
			: await cloneRepo( source );

	Logger.notice(
		`Temporary clone of ${ source } created at ${ tmpRepoPath }`
	);
```

File: tools/code-analyzer/src/lib/template-changes.ts
Comment (samueljseay) [inline tools/code-analyzer/src/lib/template-changes.ts]: This should be a const its never re-assigned.

File: tools/code-analyzer/src/lib/template-changes.ts
Comment (samueljseay) [inline tools/code-analyzer/src/lib/template-changes.ts]: TS complains since this is a string.

If it captured a string that was not a number it should probably just continue too right?

If so you could maybe do:

```
const lineNumber = parseInt(l);
if (Number.isNaN(lineNumber) || lineNumber < 4) {
...
```

File: tools/cli-core/src/git.ts
Comment (samueljseay) [inline tools/cli-core/src/git.ts]: Have you got TS setup right? This has issues.

When a function is async in JS that means it returns a promise. a promise wraps the value so the type for an async function that returns a number would be Promise<number>. but also this returns a string not a number.

**also:** What's quite nice about TS is you don't need to explicitly say what the return type is and let TS infer it. In this case if you remove `: number` from the type definition TS will see that this function returns Promise<string>.

File: tools/cli-core/src/git.ts
Comment (samueljseay) [inline tools/cli-core/src/git.ts]: Removing this `: number` allows TS to properly figure out the return type and not complain.

File: tools/release-posts/commands/release-post/release-post-release.ts
Comment (samueljseay) [inline tools/release-posts/commands/release-post/release-post-release.ts]: This should not have the ts file extension on the import, long story short its not a valid import path.

File: tools/release-posts/commands/release-post/release-post-release.ts
Comment (samueljseay) [inline tools/release-posts/commands/release-post/release-post-release.ts]: as far as I can tell you've removed the tags option so this option won't ever exist? if thats intentional this should just be a const with the default array?

File: tools/release-posts/commands/release-post/release-post-release.ts
Comment (samueljseay) [inline tools/release-posts/commands/release-post/release-post-release.ts]: The way you're declaring this and then checking it before doing stuff with options.editPost, is either too smart for TS or not typesafe.

To get rid of all the warnings I would probably just remove this variable and then where you want to check for editPost just do:

`if (options.editPost) {` since an if statement in JS is falsy for: undefined, empty string etc. Perfectly fine for this case I think.

I also think probably this option should be named like editPostId or something?

**Review by samueljseay — CHANGES_REQUESTED** [review]
I am having trouble testing the whole command locally due to internet connection so apologies for that but there are quite a lot of TS issues to resolve first so I'll retest once those are sorted.

Do you have TS properly set up in your environment because in VSCode I get lots of errors show up for these things. We're not enforcing it with any checks right now but it would be better to write the code correctly otherwise we lose all the advantages of using TypeScript.

Maybe we should punish ourselves by running TS checks for this code in PRs too though 😂

**Comment by jonathansadowski** [comment]
Thanks for having a look @samueljseay. I updated this to address all the TS issues that I saw that weren't also addressed in #36412. Once that is merged, I'm happy to incorporate those changes here and making sure that that passes. If you don't mind having another look in the mean time, or I can wait for 36412 to land.

**Review by samueljseay — DISMISSED** [review]
Apologies for the delay on the re-ping here. The new changes look good. 🚀

**Comment by jonathansadowski** [comment]
Hi @samueljseay... I had to fix a few lingering issues after #36412 was merged, otherwise merging this would've caused that lint to fail. Could you please give this another look? Thanks

**Review by samueljseay — APPROVED** [review]
Looking good! 🚀

**Pull request merged**
